packages/next/server/web/adapter.ts

```diff
@@ -26,31 +36,81 @@ export async function adapter(params: {
     },
   })
 
+  /**
+   * This allows to identify the request as a data request. The user doesn't
+   * need to know about this property neither use it. We add it for testing
+   * purposes.
+   */
+  if (buildId) {
+    Object.defineProperty(request, '__isData', {
+      enumerable: false,
+      value: true,
+    })
+  }
+
   const event = new NextFetchEvent({ request, page: params.page })
   const response = await params.handler(request, event)
 
   /**
    * For rewrites we must always include the locale in the final pathname
    * so we re-create the NextURL forcing it to include it when the it is
-   * an internal rewrite.
+   * an internal rewrite. Also we make sure the outgoing rewrite URL is
+   * a data URL if the request was a data request.
    */
-  if (response?.headers.has('x-middleware-rewrite')) {
-    const url = new NextURL(response.headers.get('x-middleware-rewrite')!, {
+  const rewrite = response?.headers.get('x-middleware-rewrite')
+  if (response && rewrite) {
+    const rewriteUrl = new NextURL(rewrite, {
       forceLocale: true,
       headers: params.request.headers,
       nextConfig: params.request.nextConfig,
     })
 
-    if (url.host === request.nextUrl.host) {
+    if (rewriteUrl.host === request.nextUrl.host) {
+      rewriteUrl.buildId = buildId || rewriteUrl.buildId
+      response.headers.set('x-middleware-rewrite', String(rewriteUrl))
+    }
+
+    /**
+     * When the request is a data request we must show if there was a rewrite
+     * with an internal header so the client knows which component to load
+     * from the data request.
+     */
+    if (buildId) {
       response.headers.set(
-        'x-middleware-rewrite',
-        String(
-          new NextURL(response.headers.get('x-middleware-rewrite')!, {
-            forceLocale: true,
-            headers: params.request.headers,
-            nextConfig: params.request.nextConfig,
-          })
-        )
+        'x-nextjs-matched-path',
```

Comment:
Is this being set for a rewrite from `next.config.js` as well? Would be good to ensure this case is being handled as well